File: audit/src/drift.js
const path = el => {
  const bits = [];
  for (let n = el; n && n.nodeType === 1 && bits.length < 4; n = n.parentElement) {
    let b = n.tagName.toLowerCase();
    if (n.id) { bits.unshift(b + '#' + n.id); break; }
    const cls = (n.getAttribute('class') || '').trim().split(/\s+/).filter(Boolean).slice(0, 2);
    if (cls.length) b += '.' + cls.join('.');
    bits.unshift(b);
  }
  return bits.join(' > ');
};

const styleOf = new Map();
const cs = el => {
  let s = styleOf.get(el);
  if (!s) { s = getComputedStyle(el); styleOf.set(el, s); }
  return s;
};

const visible = el => {
  const s = cs(el);
  if (s.display === 'none' || s.visibility === 'hidden') return false;
  if (parseFloat(s.opacity) === 0) return false;
  const r = el.getBoundingClientRect();
  return r.width > 0.5 && r.height > 0.5;
};

// Hard cap, mirroring the engine's sweep (`pixelproof-style-audit`'s
// `sweep.rs`): without one a pathological page — a form-heavy consumer screen
// with thousands of unlabelled inputs — builds a multi-megabyte JSON string
// and the CDP round-trip appears to hang. All six rules feed the single
// `component-drift` family, so one list shares the engine's per-family cap.
// Truncation is reported, never silent.
const MAX_PER_CATEGORY = 200;
const violations = [];
let truncated = false;
const push = (selector, detail) => {
  if (violations.length < MAX_PER_CATEGORY) violations.push({ selector, value: 1.0, detail });
  else truncated = true;
};

const EXEMPT_CLOSEST = '.menu, .tabs, .dropdown, .modal-backdrop, [data-ld-audit-exempt]';
// Rule 6 only. A control disabled by STATE rather than a withheld action has
// nothing to explain: the pager's current page and its boundary arrows
// ("you are here"), a busy control, and the already-selected option of a
// single-select group. Everything else that is disabled needs a reason.
const DISABLED_STATE_EXEMPT =
  '[data-pagination], [aria-current="page"], .loading, [aria-busy="true"], ' +
  '[aria-checked="true"], [aria-pressed="true"], [aria-selected="true"]';
const NON_FIELD_INPUT_TYPES = ['checkbox', 'radio', 'range', 'hidden'];

const mount = document.querySelector(OPTS.mount_selector);
const all = mount ? [mount, ...mount.querySelectorAll('*')] : [];
const els = all.filter(visible);

for (const el of els) {
  const tag = el.tagName;

  // 1. button-without-btn: a <button> whose classList lacks .btn, unless it
  // sits inside one of the exempt overlay/nav constructs, or declares itself
  // a designed unstyled action via the `data-pressable` marker — emitted by
  // the library's Pressable primitive, so intent lives in the markup rather
  // than in an exemption comment. Genuinely raw, unmarked buttons still flag.
  if (tag === 'BUTTON' && !el.classList.contains('btn')) {
    if (!el.closest(EXEMPT_CLOSEST) && !el.hasAttribute('data-pressable')) {
      push(path(el), 'button-without-btn: raw <button> lacks .btn');
    }
  }

  // 2. table-without-table-class: a <table> whose classList lacks .table.
  //
  // `.sr-only` is exempt. This rule exists to catch a table that LOOKS
  // undesigned, and a screen-reader-only table has no appearance to get
  // wrong -- every chart in this library ships one as its accessible data
  // equivalent, so without the exemption a consumer's audit reports one
  // finding per chart for markup they neither wrote nor can see. That is the
  // same failure ldui-zl58 documented for the nav rail: our own markup
  // drowning a consumer's real signal.
  if (
    tag === 'TABLE' &&
    !el.classList.contains('table') &&
    !el.classList.contains('sr-only')
  ) {
    push(path(el), 'table-without-table-class: raw <table> lacks .table');
  }

  // 3. badge-lookalike: a small, rounded, filled, text-bearing <span>/<div>
  // that never opted into .badge (or .btn, which shares the same shape).
  if (tag === 'SPAN' || tag === 'DIV') {
    const s = cs(el);
    const r = el.getBoundingClientRect();
    // Computed radius is "Npx", "N%", or "X Y" for elliptical. A percentage
    // resolves against the box, exactly as the engine's shape rule does
    // (`sweep.rs`) — `parseFloat("50%")` would otherwise read as 50px and
    // mis-flag a small round chip.
    const rawRadius = (s.borderTopLeftRadius || '').split(' ')[0];
    let btlr;
    if (rawRadius.endsWith('%')) {
      btlr = (parseFloat(rawRadius) / 100) * Math.min(r.width, r.height);
    } else {
      btlr = parseFloat(rawRadius);
    }
    if (isNaN(btlr)) btlr = 0;
    const bg = s.backgroundColor;
    const hasOwnText = Array.from(el.childNodes)
      .some(n => n.nodeType === 3 && n.textContent.trim());
    if (
      btlr >= 8 &&
      r.height > 0 && r.height <= 24 &&
      bg !== 'rgba(0, 0, 0, 0)' &&
      hasOwnText &&
      !el.classList.contains('badge') &&
      !el.classList.contains('btn')
    ) {
      push(path(el), 'badge-lookalike: pill-shaped text chip lacks .badge');
    }
  }

  // 4. input-outside-field: a text-ish <input>/<select>/<textarea> with none
  // of fieldset ancestor, wrapping label ancestor, or a matching label[for]
  // in the document.
  const isFieldCandidate =
    (tag === 'INPUT' && !NON_FIELD_INPUT_TYPES.includes((el.type || 'text').toLowerCase())) ||
    tag === 'SELECT' || tag === 'TEXTAREA';
  if (isFieldCandidate) {
    const hasFieldset = el.closest('fieldset') !== null;
    const hasWrappingLabel = el.closest('label') !== null;
    let hasLabelFor = false;
    if (el.id) {
      hasLabelFor = document.querySelector(
        'label[for="' + el.id.replace(/"/g, '\\"') + '"]'
      ) !== null;
    }
    if (!hasFieldset && !hasWrappingLabel && !hasLabelFor) {
      push(
        path(el),
        'input-outside-field: ' + tag.toLowerCase() +
        ' has no fieldset ancestor, wrapping label, or label[for]'
      );
    }
  }
}

// 5. target-too-small: WCAG 2.2 SC 2.5.8 Target Size (Minimum), level AA.
// axe-core does not run it by default (it sits outside axe's default tags),
// so a consumer's clean axe pass says nothing about it (4iiz-Office op-7ndp7:
// a primary per-row action measured 18px tall on every row).
//
// The SC is size OR spacing, and this implements both halves rather than a
// bare size check, which would flag every correctly spaced compact control:
// a target under 24x24 CSS px passes when a 24px-diameter circle centred on
// its bounding box intersects no other target and no other undersized
// target's circle. The SC's inline exception (a link in a sentence) is out of
// scope because plain links are not collected -- only buttons, `.btn`-styled
// links and ARIA buttons, which is where a dense UI's action targets live.
const TARGET_MIN = 24;
const RADIUS = TARGET_MIN / 2;
const targets = els.filter(el => {
  if (!(el.tagName === 'BUTTON' || el.matches('a.btn[href], [role="button"]'))) return false;
  if (el.disabled || el.closest('[aria-hidden="true"], [inert], [data-ld-audit-exempt]')) return false;
  const r = el.getBoundingClientRect();
  // `.sr-only` and zero-size plumbing are 1px boxes nobody can aim at.
  return r.width > 1 && r.height > 1;
}).map(el => {
  const r = el.getBoundingClientRect();
  return {
    el, r,
    cx: r.left + r.width / 2,
    cy: r.top + r.height / 2,
    small: r.width < TARGET_MIN - 0.5 || r.height < TARGET_MIN - 0.5,
  };
});
// A target nested in another target (an icon <span role="button"> inside a
// <button>) is one hit area, not two neighbours.
const related = (a, b) => a.el.contains(b.el) || b.el.contains(a.el);
const circleHitsRect = (t, r) => {
  const dx = Math.max(r.left - t.cx, 0, t.cx - r.right);
  const dy = Math.max(r.top - t.cy, 0, t.cy - r.bottom);
  return dx * dx + dy * dy < RADIUS * RADIUS;
};
for (const t of targets) {
  if (!t.small) continue;
  const crowded = targets.some(o => {
    if (o === t || related(o, t)) return false;
    if (circleHitsRect(t, o.r)) return true;
    return o.small && Math.hypot(o.cx - t.cx, o.cy - t.cy) < TARGET_MIN;
  });
  if (crowded) {
    push(
      path(t.el),
      'target-too-small: ' + Math.round(t.r.width) + 'x' + Math.round(t.r.height) +
      'px target is under 24x24 and a neighbour sits inside its 24px spacing circle (WCAG 2.5.8)'
    );
  }
}

// 6. disabled-without-reason (ldui-p82h): a disabled control -- a native
// `<button disabled>`, anything carrying daisyUI's `.btn-disabled`, or an
// ARIA button with `aria-disabled="true"` -- must still have an accessible
// name (a disabled control is still announced, and an icon-only one with no
// name is an unlabelled dead end) AND an `aria-describedby` that resolves to
// non-empty text saying WHY it is disabled. The library's `Button` mints
// both from its `disabled_reason` prop, and stamps
// `data-disabled-without-reason="true"` when it was rendered `disabled=true`
// without a reason, so that marker is reported verbatim in the finding
// wherever it appears -- it names the fix (pass `disabled_reason`).
//
// The name is the accname computation's first three steps, not the whole
// algorithm: `aria-label`, then `aria-labelledby`, then the element's own
// text with `aria-hidden` subtrees excluded (the Button's reason hint is one,
// deliberately, so a reason can never masquerade as a name), then `title`.
// A description resolves through `aria-describedby` only; `title` is not
// accepted as a description because it is not announced on every platform.
const ownText = el => {
  let out = '';
  const walk = n => {
    for (const c of n.childNodes) {
      if (c.nodeType === 3) out += c.textContent;
      else if (c.nodeType === 1 && c.getAttribute('aria-hidden') !== 'true') walk(c);
    }
  };
  walk(el);
  return out.replace(/\s+/g, ' ').trim();
};
const idrefsText = (el, attr) => (el.getAttribute(attr) || '')
  .split(/\s+/)
  .filter(Boolean)
  .map(id => { const t = document.getElementById(id); return t ? t.textContent.trim() : ''; })
  .filter(Boolean)
  .join(' ');
for (const el of els) {
  const isDisabled =
    (el.tagName === 'BUTTON' && el.disabled) ||
    el.classList.contains('btn-disabled') ||
    (el.matches('[role="button"]') && el.getAttribute('aria-disabled') === 'true');
  if (!isDisabled) continue;
  if (el.closest(EXEMPT_CLOSEST) || el.closest('[aria-hidden="true"], [inert]')) continue;
  if (el.closest(DISABLED_STATE_EXEMPT)) continue;
  const name =
    (el.getAttribute('aria-label') || '').trim() ||
    idrefsText(el, 'aria-labelledby') ||
    ownText(el) ||
    (el.getAttribute('title') || '').trim();
  const reason = idrefsText(el, 'aria-describedby');
  const marker = el.hasAttribute('data-disabled-without-reason')
    ? ' [data-disabled-without-reason]'
    : '';
  if (!name) {
    push(path(el), 'disabled-without-reason: disabled control has no accessible name' + marker);
  } else if (!reason) {
    push(
      path(el),
      'disabled-without-reason: disabled control has no aria-describedby resolving to non-empty text' + marker
    );
  }
}

return JSON.stringify({ violations, scanned: els.length, truncated });

File: audit/src/drift.rs
//! daisyUI component-drift heuristics: a second, small in-page sweep separate
//! from the engine's generic style sweep. The engine (`pixelproof-style-audit`)
//! knows no framework, only values; these six rules are daisyUI and WCAG
//! knowledge —
//! "a raw `<button>` without `.btn`", "a raw `<table>` without `.table`",
//! "a disabled control that never says why" —
//! so they live here rather than in the engine.
//!
//! Follows the engine's `sweep_js` pattern (see
//! `pixelproof_style_audit::sweep_js` / `sweep.rs`): a self-contained JS IIFE
//! with the mount selector embedded as a JSON constant, walking the visible
//! elements under the mount subtree and returning a JSON-stringified
//! `{ violations, scanned, truncated }` object — including the engine's hard
//! per-family cap, so a pathological page cannot build a multi-megabyte JSON
//! string and hang the CDP round-trip. [`crate::page::audit_page`] merges this
//! report's violations into the engine's
//! [`pixelproof_style_audit::AuditReport`].

use pixelproof_web::Harness;
use serde::Serialize;

/// Raw result of the drift sweep on one page.
#[derive(Debug, serde::Deserialize)]
pub struct DriftReport {
    pub violations: Vec<pixelproof_style_audit::Violation>,
    /// Elements considered — 0 means the mount selector matched nothing.
    pub scanned: usize,
    /// Whether the sweep's per-family cap was hit, so `violations` is a floor
    /// rather than a total. Mirrors `AuditReport::truncated`; `serde(default)`
    /// so an older sweep body still deserializes.
    #[serde(default)]
    pub truncated: bool,
}

/// Surface-shape options for the drift sweep — currently just where to look.
/// Kept as its own JSON-embedded struct (rather than a bare string) to mirror
/// the engine's `OPTS` pattern and leave room for future exemption knobs.
#[derive(Debug, Clone, Serialize)]
struct DriftOpts {
    mount_selector: String,
}

/// Generate the drift sweep IIFE for one mount selector (embedded as a JSON
/// constant, mirroring the engine's `sweep_js` pattern).
pub fn drift_js(mount_selector: &str) -> String {
    let opts = DriftOpts {
        mount_selector: mount_selector.to_string(),
    };
    let opts_json = serde_json::to_string(&opts).expect("drift options serialize");
    format!("(() => {{\nconst OPTS = {opts_json};\n{DRIFT_BODY}\n}})()")
}

/// The static drift-sweep body, authored in `drift.js` for readability (a
/// plain `.js` file, not a Rust string literal) and pulled in at compile
/// time. Reads only `OPTS`. Its braces are safe because it enters the
/// `format!` wrapper as an argument, not as format text.
const DRIFT_BODY: &str = include_str!("drift.js");

/// Evaluate the drift sweep on the currently-loaded page.
///
/// Mirrors the engine's `web::run_sweep` mechanics: evaluate the generated
/// JS, expect a JSON string back (CDP value marshalling flattens nested
/// arrays inconsistently across driver versions — a string round-trips
/// identically everywhere), `serde_json::from_str`, map errors to a
/// descriptive `String` (this layer does not own the engine's error enum).
pub async fn run_drift(h: &Harness, mount_selector: &str) -> Result<DriftReport, String> {
    let js = drift_js(mount_selector);
    let raw: String = h
        .page()
        .evaluate(js)
        .await
        .map_err(|e| format!("drift sweep did not evaluate: {e}"))?
        .into_value()
        .map_err(|e| format!("drift sweep did not evaluate to a string: {e}"))?;
    serde_json::from_str(&raw)
        .map_err(|error| format!("drift sweep returned unparseable JSON ({error}): {raw}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn drift_js_embeds_the_mount_selector_as_json() {
        let js = drift_js("main");
        assert!(js.contains("\"main\""));
        // Every rule id must appear in the generated source, so a rule
        // cannot be silently dropped.
        for rule in [
            "button-without-btn",
            "table-without-table-class",
            "badge-lookalike",
            "input-outside-field",
            "target-too-small",
            "disabled-without-reason",
        ] {
            assert!(js.contains(rule), "rule {rule} missing from generated JS");
        }
    }

    /// ldui-p82h: the a11y half of `disabled_reason`. The rule must check
    /// BOTH halves -- an accessible name and an `aria-describedby` that
    /// resolves to text -- and must report the Button's own
    /// `data-disabled-without-reason` marker so the finding names the fix.
    #[test]
    fn disabled_without_reason_rule_checks_name_and_description_and_reports_the_marker() {
        let js = drift_js("main");
        assert!(
            js.contains("disabled-without-reason"),
            "the rule is embedded"
        );
        assert!(
            js.contains("el.tagName === 'BUTTON' && el.disabled"),
            "a native disabled button is a disabled control"
        );
        assert!(
            js.contains("el.classList.contains('btn-disabled')"),
            "daisyUI's visual-only disabled class is a disabled control too"
        );
        assert!(
            js.contains("idrefsText(el, 'aria-describedby')"),
            "the reason must be resolved through aria-describedby idrefs, not assumed from the attribute's presence"
        );
        assert!(
            js.contains("disabled control has no accessible name"),
            "a nameless disabled control is the first finding"
        );
        assert!(
            js.contains("has no aria-describedby resolving to non-empty text"),
            "a described-by that resolves to nothing is the second finding"
        );
        assert!(
            js.contains("el.hasAttribute('data-disabled-without-reason')"),
            "the Button's marker must be surfaced in the finding"
        );
    }

    /// The negative control for the name computation: the Button's reason
    /// hint is an `aria-hidden` descendant, so the rule's own-text walk must
    /// skip `aria-hidden` subtrees -- otherwise a reason would count as a
    /// name and an icon-only disabled button with no label would pass.
    #[test]
    fn disabled_without_reason_name_excludes_aria_hidden_descendants() {
        let js = drift_js("main");
        assert!(
            js.contains("c.getAttribute('aria-hidden') !== 'true') walk(c)"),
            "the own-text walk must not descend into aria-hidden subtrees"
        );
        assert!(
            js.contains("if (!isDisabled) continue;"),
            "an enabled control is never a finding of this rule (the rule's own negative control)"
        );
    }

    /// State-disabled controls are not findings: a pager's current page and
    /// boundary arrows, a busy control, and a single-select group's chosen
    /// option carry no reason because there is nothing to explain. The
    /// exemption is a closed selector list, applied only in rule 6, so a
    /// genuinely reasonless action can never hide behind it.
    #[test]
    fn disabled_without_reason_exempts_state_disabled_controls_only() {
        let js = drift_js("main");
        for hook in [
            "[data-pagination]",
            "[aria-current=\"page\"]",
            ".loading",
            "[aria-busy=\"true\"]",
            "[aria-checked=\"true\"]",
            "[aria-pressed=\"true\"]",
            "[aria-selected=\"true\"]",
        ] {
            assert!(
                js.contains(hook),
                "state exemption {hook} missing from DISABLED_STATE_EXEMPT"
            );
        }
        assert!(
            js.contains("if (el.closest(DISABLED_STATE_EXEMPT)) continue;"),
            "the exemption is applied by ancestry inside rule 6"
        );
        assert_eq!(
            js.matches("DISABLED_STATE_EXEMPT").count(),
            2,
            "declared once, used once -- no other rule may borrow the state exemption"
        );
    }

    /// The library's `Pressable` primitive emits `data-pressable="true"` as
    /// its auditable marker; the button rule must recognize it so a designed
    /// unstyled action never counts as drift (and the counts drop honestly,
    /// not via exemption comments). Genuinely raw buttons keep flagging.
    #[test]
    fn target_size_rule_implements_both_halves_of_sc_2_5_8() {
        let js = drift_js("main");
        assert!(js.contains("target-too-small"), "the rule is embedded");
        assert!(
            js.contains("const TARGET_MIN = 24;"),
            "the AA minimum is 24px"
        );
        assert!(
            js.contains("circleHitsRect"),
            "the spacing exception is implemented, not a bare size check"
        );
    }

    #[test]
    fn button_rule_recognizes_the_pressable_marker() {
        let js = drift_js("main");
        assert!(
            js.contains("hasAttribute('data-pressable')"),
            "button-without-btn must accept the Pressable marker attribute"
        );
    }

    #[test]
    fn drift_report_deserializes_the_sweep_shape() {
        let r: DriftReport = serde_json::from_str(
            r#"{"violations":[{"selector":"button#save","value":1.0,"detail":"button-without-btn: raw button lacks .btn"}],"scanned":42,"truncated":false}"#,
        )
        .unwrap();
        assert_eq!(r.violations.len(), 1);
        assert_eq!(r.scanned, 42);
        assert!(!r.truncated);
    }

    #[test]
    fn drift_report_carries_truncation() {
        let r: DriftReport =
            serde_json::from_str(r#"{"violations":[],"scanned":9000,"truncated":true}"#).unwrap();
        assert!(
            r.truncated,
            "a capped sweep must report that its counts are a floor"
        );
        // And an older sweep body that predates the flag still parses.
        let r: DriftReport = serde_json::from_str(r#"{"violations":[],"scanned":9000}"#).unwrap();
        assert!(!r.truncated);
    }

    /// The cap is the whole point of mirroring the engine here: without it a
    /// form-heavy consumer screen builds a multi-megabyte JSON string and the
    /// CDP round-trip appears to hang.
    #[test]
    fn generated_drift_sweep_caps_its_violation_list() {
        let js = drift_js("main");
        assert!(
            js.contains("MAX_PER_CATEGORY = 200"),
            "drift sweep must carry the engine's per-family cap"
        );
        assert!(
            js.contains("truncated = true"),
            "hitting the cap must latch the truncation flag"
        );
        assert!(
            js.contains("scanned: els.length, truncated"),
            "the sweep must return the truncation flag, not just record it"
        );
    }

    /// A percentage radius resolves against the box, exactly as the engine's
    /// shape rule does — `parseFloat("50%")` would read as 50px and mis-flag
    /// a small round chip as a badge lookalike.
    #[test]
    fn badge_lookalike_resolves_percentage_radius_against_the_box() {
        let js = drift_js("main");
        assert!(
            js.contains("rawRadius.endsWith('%')"),
            "percentage radius must be detected"
        );
        assert!(
            js.contains("(parseFloat(rawRadius) / 100) * Math.min(r.width, r.height)"),
            "percentage radius must resolve against the shorter box side"
        );
    }

    #[test]
    fn generated_drift_sweep_is_a_single_invoked_iife() {
        let js = drift_js("main");
        let js = js.trim();
        assert!(js.starts_with("(() => {"), "must be an IIFE");
        assert!(js.ends_with("})()"), "must be invoked");
    }

    #[test]
    fn custom_mount_selector_is_respected() {
        let js = drift_js("#app");
        assert!(js.contains(r##""mount_selector":"#app""##));
    }
}
